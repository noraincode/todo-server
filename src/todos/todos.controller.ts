import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  UsePipes,
  ValidationPipe,
  UseInterceptors,
} from '@nestjs/common';
import { TodosService } from './todos.service';
import { CreateTodoDto } from './dto/create-todo.dto';
import { UpdateTodoDto } from './dto/update-todo.dto';
import { ResponseInterceptor } from 'src/nest/interceptors/response.interceptor';

@Controller('todos')
@UseInterceptors(ResponseInterceptor)
export class TodosController {
  constructor(private readonly todosService: TodosService) {}

  @Post()
  @UsePipes(new ValidationPipe({ errorHttpStatusCode: 422 }))
  create(@Body() createTodoDto: CreateTodoDto) {
    return this.todosService.create(createTodoDto);
  }

  @Get()
  findAll() {
    return this.todosService.findAll();
  }

  @Patch('complete-batches')
  completeAll() {
    return this.todosService.completeAll();
  }

  @Delete('completed')
  async removeCompletedTodos() {
    await this.todosService.removeCompletedTodos();
    return { message: 'Completed todos have been deleted.' };
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.todosService.findOne(id);
  }

  @Patch(':id')
  update(@Param('id') id: string, @Body() updateTodoDto: UpdateTodoDto) {
    return this.todosService.update(id, updateTodoDto);
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.todosService.remove(id);
  }
}
