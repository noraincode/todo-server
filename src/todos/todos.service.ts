import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Todos } from './entities/todo.entity';
import { CreateTodoDto } from './dto/create-todo.dto';
import { UpdateTodoDto } from './dto/update-todo.dto';

@Injectable()
export class TodosService {
  constructor(
    @InjectRepository(Todos)
    private todosRepository: Repository<Todos>,
  ) {}

  create(createTodoDto: CreateTodoDto) {
    const todo = this.todosRepository.create(createTodoDto);
    return this.todosRepository.save(todo);
  }

  findAll() {
    return this.todosRepository.find();
  }

  async findOne(id: string) {
    const todo = await this.todosRepository.findOne({ where: { id } });
    if (!todo) {
      throw new NotFoundException('Todo not found');
    }

    return todo;
  }

  async update(id: string, updateTodoDto: UpdateTodoDto) {
    const todo = await this.todosRepository.findOne({ where: { id } });
    if (!todo) {
      throw new NotFoundException('Todo not found');
    }
    if (updateTodoDto.completed) {
      updateTodoDto.completed_at = new Date();
    }
    this.todosRepository.merge(todo, updateTodoDto);
    return this.todosRepository.save(todo);
  }

  async remove(id: string) {
    const todo = await this.todosRepository.findOne({ where: { id } });
    if (!todo) {
      throw new NotFoundException('Todo not found');
    }

    return this.todosRepository.delete(id);
  }

  async completeAll() {
    const todos = await this.todosRepository.find({
      where: { completed: false },
    });
    if (todos.length === 0) {
      throw new NotFoundException('No uncompleted todos found');
    }

    const now_date = new Date();

    const todos_to_complete = todos.map((todo) => {
      return {
        ...todo,
        completed: true,
        completed_at: now_date,
      };
    });

    return this.todosRepository.save(todos_to_complete);
  }

  async removeCompletedTodos() {
    const completed_todos = await this.todosRepository.find({
      where: { completed: true },
    });
    if (completed_todos.length === 0) {
      throw new NotFoundException('No completed todos found');
    }

    return this.todosRepository.delete({ completed: true });
  }
}
