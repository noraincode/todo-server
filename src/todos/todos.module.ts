import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { TodosService } from './todos.service';
import { TodosController } from './todos.controller';
import { Todos } from './entities/todo.entity';

@Module({
  imports: [TypeOrmModule.forFeature([Todos])],
  providers: [TodosService],
  controllers: [TodosController],
})
export class TodosModule {}
