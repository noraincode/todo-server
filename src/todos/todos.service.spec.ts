/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-nocheck

import { Test, TestingModule } from '@nestjs/testing';
import { TodosService } from './todos.service';
import { Todos } from './entities/todo.entity';
import { Repository } from 'typeorm/repository/Repository';
import { getRepositoryToken } from '@nestjs/typeorm';
import MockDate from 'mockdate';

describe('TodosService', () => {
  let service: TodosService;
  let mockRepository: Partial<Repository<Todos>>;

  beforeEach(async () => {
    jest.resetAllMocks();

    mockRepository = {
      find: jest.fn(),
      findOne: jest.fn(),
      create: jest.fn(),
      save: jest.fn(),
      merge: jest.fn(),
      delete: jest.fn(),
      update: jest.fn(),
    };
    MockDate.set('2024-02-01');

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        TodosService,
        {
          provide: getRepositoryToken(Todos),
          useValue: mockRepository,
        },
      ],
    }).compile();

    service = module.get<TodosService>(TodosService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('findAll', () => {
    it('should return an array of todos', async () => {
      const test_todos = [
        { id: 'mock_id', title: 'Test Todo', isCompleted: false },
      ];
      mockRepository.find.mockReturnValue(test_todos);

      expect(await service.findAll()).toBe(test_todos);
    });
  });

  describe('create', () => {
    it('should add a new todo and return it', async () => {
      const newTodo = { title: 'New Todo' };
      const savedTodo = { id: 'mock_id', ...newTodo };
      mockRepository.create.mockReturnValue(newTodo);
      mockRepository.save.mockResolvedValue(savedTodo);

      expect(await service.create(newTodo)).toEqual(savedTodo);
      expect(mockRepository.create.mock.calls).toMatchSnapshot(
        'calls to create',
      );
      expect(mockRepository.save.mock.calls).toMatchSnapshot('calls to save');
    });
  });

  describe('findOne', () => {
    it('should return a single todo item', async () => {
      const todoId = 'mock_id';
      const foundTodo = { id: todoId, title: 'Found Todo' };
      mockRepository.findOne.mockResolvedValue(foundTodo);

      expect(await service.findOne(todoId)).toEqual(foundTodo);
      expect(mockRepository.findOne.mock.calls).toMatchSnapshot(
        'calls to findOne',
      );
    });
  });

  describe('update', () => {
    it('should update a todo and return the updated todo', async () => {
      const todoId = 'mock_id';
      const updateData = { title: 'Updated Todo' };
      const updatedTodo = { id: todoId, ...updateData };
      mockRepository.findOne.mockResolvedValue(updatedTodo);
      mockRepository.save.mockResolvedValue(updatedTodo);

      expect(await service.update(todoId, updateData)).toEqual(updatedTodo);
      expect(mockRepository.findOne.mock.calls).toMatchSnapshot(
        'calls to findOne',
      );
      expect(mockRepository.save.mock.calls).toMatchSnapshot('calls to save');
    });
  });

  describe('completeAll', () => {
    it('should mark all todos as completed', async () => {
      const test_todos = [
        { id: 'mock_id', title: 'Test Todo', completed: false },
      ];
      mockRepository.find.mockReturnValue(test_todos);
      mockRepository.save.mockResolvedValue({ affected: 1 });

      expect(await service.completeAll()).toEqual({ affected: 1 });
      expect(mockRepository.save.mock.calls).toMatchSnapshot('calls to save');
    });
  });

  describe('removeCompletedTodos', () => {
    it('should remove all completed todos', async () => {
      const test_todos = [
        { id: 'mock_id', title: 'Test Todo', completed: true },
      ];
      mockRepository.find.mockReturnValue(test_todos);
      mockRepository.delete.mockResolvedValue({ affected: 1 }); // 假设删除操作返回受影响的行数

      expect(await service.removeCompletedTodos()).toEqual({ affected: 1 });
      expect(mockRepository.delete.mock.calls).toMatchSnapshot(
        'calls to delete',
      );
    });
  });
});
