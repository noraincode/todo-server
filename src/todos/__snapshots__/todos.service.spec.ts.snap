// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosService completeAll should mark all todos as completed: calls to save 1`] = `
[
  [
    [
      {
        "completed": true,
        "completed_at": 2024-02-01T00:00:00.000Z,
        "id": "mock_id",
        "title": "Test Todo",
      },
    ],
  ],
]
`;

exports[`TodosService create should add a new todo and return it: calls to create 1`] = `
[
  [
    {
      "title": "New Todo",
    },
  ],
]
`;

exports[`TodosService create should add a new todo and return it: calls to save 1`] = `
[
  [
    {
      "title": "New Todo",
    },
  ],
]
`;

exports[`TodosService findOne should return a single todo item: calls to findOne 1`] = `
[
  [
    {
      "where": {
        "id": "mock_id",
      },
    },
  ],
]
`;

exports[`TodosService removeCompletedTodos should remove all completed todos: calls to delete 1`] = `
[
  [
    {
      "completed": true,
    },
  ],
]
`;

exports[`TodosService update should update a todo and return the updated todo: calls to findOne 1`] = `
[
  [
    {
      "where": {
        "id": "mock_id",
      },
    },
  ],
]
`;

exports[`TodosService update should update a todo and return the updated todo: calls to save 1`] = `
[
  [
    {
      "id": "mock_id",
      "title": "Updated Todo",
    },
  ],
]
`;
